aten/src/ATen/native/cuda/CusparseLtKernels.cu

```diff
@@ -0,0 +1,349 @@
+    return 8; // 1000
+  }
+  if (pos_tuple == std::make_tuple(0, 1, 1, 0)) {
+    return 9; // 1001
+  }
+  if (pos_tuple == std::make_tuple(1, 0, 0, 1)) {
+    return 12; // 1100
+  }
+  if (pos_tuple == std::make_tuple(0, 1, 0, 1)) {
+    return 13; // 1101
+  }
+  if (pos_tuple == std::make_tuple(0, 0, 1, 1)) {
+    return 14; // 1110
+  }
+  TORCH_CHECK(
+      false,
+      "Unsupported mask configuration: ",
+      pos0,
+      pos1,
+      pos2,
+      pos3,
+      ". Please follow 2 by 4 pattern.");
+  return 0;
+}
+
+// Based on
+// https://github.com/NVIDIA/cutlass/blob/8b42e751c63ba219755c8ed91af5f6ec1ecc1ee6/tools/util/include/cutlass/util/host_reorder.h#L86
+// Original note:
+// This is needed for the sparse tensor core kernels.  The purpose
+// is to use ldmatrix to load from shared memory to the register file.
+void reorder_meta(int16_t* dest,
+                  const int16_t* src,
+                  int64_t length_m,
+                  int64_t meta_size_1) {
+  for (int m = 0; m < length_m; m++) {
+    for (int k = 0; k < meta_size_1; k++) {
+      // NOTE: Kept full code for future extensibility
+      // First reorder the rows.
+      int group = (sizeof(int16_t) == 2) ? 32 : 16;
+      int interweave = (sizeof(int16_t) == 2) ? 4 : 2;
+
+      int dest_row = m / group * group + (m % 8) * interweave + (m % group) / 8;
+      int dest_col = k;
+
+      // Next swizzle the 2x2 blocks from Z to N.
+      if (((dest_row % 2) == 0) && ((dest_col % 2) == 1)) {
+        ++dest_row;
+        --dest_col;
+      } else if (((dest_row % 2) == 1) && ((dest_col % 2) == 0)) {
+        --dest_row;
+        ++dest_col;
+      }
+
+      dest[dest_row * meta_size_1 + dest_col] = src[m * meta_size_1 + k];
+    }
+  }
+}
+
+Tensor _cusparselt_create_meta(const Tensor& mask) {
+  const int64_t length_m = mask.size(0);
+  const int64_t length_k = mask.size(1);
+  TORCH_CHECK(
+      length_k % 16 == 0, "Expected size(1) of mask to be divisible by 16.");
+  TORCH_CHECK(mask.is_contiguous(), "Expected mask to be contiguous.");
+  TORCH_CHECK(mask.dtype() == at::kBool, "Expected mask to be of dtype bool.");
+  auto options = mask.options();
+  options = options.dtype(at::kShort);
+  auto result = at::empty({length_m, length_k / 16}, options);
+  auto result_reordered = at::empty({length_m, length_k / 16}, options);
+  const bool* mask_ptr = mask.data_ptr<bool>();
+  int16_t* result_ptr = result.data_ptr<int16_t>();
+  for (int64_t i = 0; i < length_m; i++) {
+    for (int64_t j = 0; j < length_k; j += 16) {
+      result_ptr[i * (length_k / 16) + (j / 16)] = 0;
+      uint16_t result_val = 0;
+      for (int64_t k = 0; k < 4; k++) {
+        bool pos0 = mask_ptr[i * length_k + j + k * 4];
+        bool pos1 = mask_ptr[i * length_k + j + k * 4 + 1];
+        bool pos2 = mask_ptr[i * length_k + j + k * 4 + 2];
+        bool pos3 = mask_ptr[i * length_k + j + k * 4 + 3];
+        uint16_t meta_val = _mask_to_meta(pos0, pos1, pos2, pos3);
```

Comment:
I believe there exists another approach to computing `meta_val` that might be more efficient (it avoids `std::tuple`):
```c++
const int32_t pos = reinterpret_cast<const int32_t>(mask_ptr + i * length_k + j + k * 4)[0];
uint16_t meta_val = 0;
switch(pos) {
  case 257: meta_val = 4; break;
  case 65537: meta_val = 8; break;
  case 65792: meta_val = 9; break;
  case 16777217: meta_val = 12; break;
  case 16777472: meta_val = 13; break;
  case 16842752: meta_val = 14; break;
  default: break;
}
TORCH_CHECK(meta_val > 0, "Unsupported mask configuration: ...");
```
(untested, assuming torch.bool is equivalent to torch.uint8).

In fact, with some effort, it should be possible to eliminate the loop over `k` altogether and replace it with two switch statements to compute `result_val`. This optimization would make sense after having solid tests of the feature in place. So, feel free to ignore this comment.